lnrpc/routerrpc/forward_interceptor.go

```diff
@@ -0,0 +1,224 @@
+package routerrpc
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/htlcswitch"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// ErrFwdNotExists is an error returned when the caller tries to resolve
+	// a forward that doesn't exist anymore.
+	ErrFwdNotExists = errors.New("forward does not exist")
+
+	// ErrMissingPreimage is an error returned when the caller tries to settle
+	// a forward and doesn't provide a preimage.
+	ErrMissingPreimage = errors.New("missing preimage")
+)
+
+// forwardInterceptor is a helper struct that handles the lifecycle of an rpc
+// interceptor streaming session.
+// It is created when the stream opens and disconnects when the stream closes.
+type forwardInterceptor struct {
+
+	// server is the Server reference
+	server *Server
+
+	// holdForwards is a map of current hold forwards and their corresponding
+	// ForwardResolver.
+	holdForwards map[channeldb.CircuitKey]htlcswitch.InterceptedForward
+
+	// stream is the bidirectional RPC stream
+	stream Router_HtlcInterceptorServer
+
+	// quit is a channel that is closed when this forwardInterceptor is shutting
+	// down.
+	quit chan struct{}
+
+	// intercepted is where we stream all intercepted packets coming from
+	// the switch.
+	intercepted chan htlcswitch.InterceptedForward
+
+	wg sync.WaitGroup
+}
+
+// newforwardInterceptor creates a new forwardInterceptor.
+func newforwardInterceptor(server *Server, stream Router_HtlcInterceptorServer) *forwardInterceptor {
+	return &forwardInterceptor{
+		server: server,
+		stream: stream,
+		holdForwards: make(
+			map[channeldb.CircuitKey]htlcswitch.InterceptedForward),
+		quit:        make(chan struct{}),
+		intercepted: make(chan htlcswitch.InterceptedForward),
+	}
+}
+
+// run sends the intercepted packets to the client and receives the
+// corersponding responses. On one hand it regsitered itself as an interceptor
+// that receives the switch packets and on the other hand launches a go routine
+// to read from the client stream.
+// To coordinate all this and make sure it is safe for concurrent access all
+// packets are sent to the main where they are handled.
+func (r *forwardInterceptor) run() error {
+	// make sure we disconnect and resolves all remaining packets if any.
+	defer r.onDisconnect()
+
+	// Register our interceptor so we receive all forwarded packets.
+	interceptableForwarder := r.server.cfg.RouterBackend.InterceptableForwarder
+	interceptableForwarder.SetInterceptor(r.onIntercept)
```

Comment:
It is nice to have the set and clear on the same level. Maybe take the clear out of disconnect and do an additional defer call?

Reply:
Still interested to hear your thought on this. It is a non-blocking suggestion though

Reply:
> Still interested to hear your thought on this. It is a non-blocking suggestion though

I agree it is better so I changed it.